subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskContainer.java

```diff
@@ -475,26 +472,50 @@ public <T extends Task> T createWithoutConstructor(String name, Class<T> type, l
     }
 
     @Override
-    public Task findByPath(String path) {
-        Path.validatePath(path);
-        if (!path.contains(Project.PATH_SEPARATOR)) {
-            return findByName(path);
+    public @Nullable Task findByPath(String pathStr) {
+        Path path = Path.path(pathStr);
+        if (!path.isAbsolute() && path.getParent() == null) {
```

Comment:
💭 This condition appears in multiple places and is not trivial. Can we provide an intention-revealing method for this? Would also be nice to avoid a Path allocation for the parent, since it's immediately discarded.

Reply:
I've replaced this with `!path.isAbsolute() && path.segmentCount() == 1`, which avoids the allocation. 

We could choose to add a `boolean Path.isName()` method which returns true if this path is not absolute and has no parent.